### PR TITLE
Shocks propagate to people you are pulling and pulled by.

### PR DESCRIPTION
**Pull request by kriskog**
## About The Pull Request
What it says on the tin. If you for example run into a shocked door while being pulled by or pulling someone, they get shocked for 75% of the damage. Shock immune applies as normal down the chain.
Kept it to /mob/living/carbon only, simplemobs got it rough enough as is.

Also changed grounded nerves to use TRAIT_SHOCKIMMUNE instead of altering its species siemens_coeff. 

## Why It's Good For The Game
Adds a bit of realism.

## Changelog
:cl: Skoglol
add: Shocks now propagate to people you are pulling or pulled by.
code: Grounded nerves now uses the shock immune trait.
/:cl:

### CODE REVIEW

**Comment by Teepeepee** [comment]
time to make conga lines and run into a grille

**Comment by subject217** [comment]
How will this work with defib drag shocking?

**Comment by kriskog** [comment]
Hm. Fair point, they do get shocked currently. I'll sleep on it.

I'll be adding some text to this as well, akin to the one you get from defibbing someone you are pulling.
image

**Comment by WJohn** [comment]
CONGA TRAIN CHAIN LIGHTNING

I AM DOWN FOR THIS

**Comment by MrPerson** [comment]
There's an easy way to solve the problem of defibs shocking people unintentionally. Just shout "clear" before using them. Then when they get shocked, you can go "I yelled clear!". Hey, it works for the janitor!

**Comment by kriskog** [comment]
Yes, that is true. And while tempting, in the case of defibbing I don't want you to be able to get shocked by a shock originating from yourself.

**Comment by subject217** [comment]
The point is moreso that it makes zero sense for a "defibrillatory" shock to suddenly hurt the person being shocked because it arced to the person holding them and then back to them. A better solution would be to change how defib shocks work as like an extension of this system rather than being at ends with it.

**Comment by MrPerson** [comment]
Maybe the shock needs a list of already shocked things passed around as an arg then. Then you could also handle shocks up the chain (ie if you get shocked and you're being pulled, shock them too) without creating an infinite loop.

**Comment by kriskog** [comment]
Having shocks go both ways is an easy task, it is something I have considered. As for infinite loops, you'll drop whatever you are pulling when getting shocked, so the chain breaks automagically.

Rewriting large parts of electrocute_act or adding args to avoid rare edge cases wasn't my intention for this, I am sure I'll find a simple solution when I take a new look at it. Tomorrow, probably.

**Comment by MrPerson** [comment]
Yeah I wasn't going to say anything, but w/e. Other changes are out of scope. If you want to do them, great, but this PR is good as-is.

**Comment by Time-Green** [comment]
If a group of pulling people get hit by lightning are they just all dead?

**Comment by GuyonBroadway** [comment]
So its theoretically possible for a group of four people to grab each other in a loop

1 grabs 2 grabs 3 grabs 4 grabs 1

If one of those people gets a shock... What happens?

**Comment by kriskog** [comment]
Whenever you get shocked, you are stunned. This makes you drop people you are pulling, so the loop breaks. Even if it didn't, the shock loses 25% potency every time it jumps, ie 30 shock damage will only jump 12 times with no protection.

**Comment by GuyonBroadway** [comment]
While that's sensible its also mildly disappointing.

**Comment by knacker48** [comment]
Those this affect the shock touch mutation?

**Comment by ATH1909** [comment]
Why shouldn't this be made to work with simplemobs/non-carbons?

**Comment by Goodfishie** [comment]
Just a note, defib shocks aren't really all that potent anyway, there are some healthcare practicioners who prefer to shock WHILE doing CPR to keep blood flow going, it doesn't really even hurt much

**Comment by SouDescolado** [comment]
Could wet floor remove the 25% reduction? Or maybe lower, like 10% reduction if the floor is wet? (First gets 100%, second 90%, etc)

**Comment by Barhandar** [comment]
Magic electricity prefers going through people than through the very conductive flooring.

**Comment by DaxDupont** [comment]
Won't this eliminate being able to escape if the idiot dragging you in cuffs walks into a grille?

**Comment by kriskog** [comment]
> Why shouldn't this be made to work with simplemobs/non-carbons?
They do not have a shock stun, and I dont believe they drop whoever they are dragging. Which means a simplemob dragging you and suiciding into a grille could conceivably instakill you. There is also no visual feedback of  simplemobs getting shocked. 

>Won't this eliminate being able to escape if the idiot dragging you in cuffs walks into a grille?
Eliminate? No. It will stun you both if you aren't wearing insuls, but he will drop you - use that to get away I guess.

>Those this affect the shock touch mutation?
Any shock that uses electrocute_act() is affected by this. Immunities apply as normal.

**Comment by ExcessiveUseOfCobblestone** [comment]
lol

File: code/game/objects/items/defib.dm
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/defib.dm]: These should be covered by carbon `electrocute_act`

File: code/modules/mob/living/carbon/human/human_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/human/human_defense.dm]: These should be covered by carbon `electrocute_act`

File: code/modules/mob/living/carbon/human/human_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/human/human_defense.dm]: These should be covered by carbon `electrocute_act`

**Review by ExcessiveUseOfCobblestone — COMMENTED** [review]
Double check behind me https://github.com/tgstation/tgstation/blob/180c87f8e0da2ae83bed8c0df98d0513fc48778c/code/modules/mob/living/carbon/carbon_defense.dm#L226

https://github.com/tgstation/tgstation/blob/180c87f8e0da2ae83bed8c0df98d0513fc48778c/code/modules/mob/living/carbon/human/human_defense.dm#L462

File: code/game/objects/items/defib.dm
Comment (kriskog) [reply code/game/objects/items/defib.dm]: This one is an exception to the electrocute_act rule. It needs the extra checks to stop pulling before getting shocked, to prevent shocking the defibee.

Comment (ExcessiveUseOfCobblestone) [reply code/game/objects/items/defib.dm]: You right

**Comment by ExcessiveUseOfCobblestone** [comment]
Another option is adding a `transferable = TRUE` arg to electrocute_act that handles whether or not it can transfer to pull/pulledby then set it to `FALSE` on the pull/pulledby's `electrocute_act`. Reasoning being if it's not the type that stuns I don't see why it needs to stop pulling and I guess realism too.

**Comment by kriskog** [comment]
@ExcessiveUseOfCobblestone I was already working on changing the logic of this. By moving it to carbon, it allows for monkeys and xenos to be affected by this. In addition, I can run the code inbetween deciding damage is dealt and the stun, which lets me use electrocute_act()'s checks for immunity instead of the additional checks.

Working on that it also occured to me that I never passed along  the other args to whoever got shocked down the chain, so you could technically have a chain of people shocked (and actually hurt) by an illusion :^)

**Comment by PKPenguin321** [comment]
>Those this affect the shock touch mutation?

Yes
t. coder of mutation

**Comment by ATH1909** [comment]
> Maybe the shock needs a list of already shocked things (EDIT: or excluded things, like the mob holding the defib) passed around as an arg then. Then you could also handle shocks up the chain (ie if you get shocked and you're being pulled, shock them too) without creating an infinite loop.

This sounds like the best solution, and I think that this would solve the "simplemob suiciding into a grill" issue that's keeping us from letting simplemobs join in on the fun.

**Comment by ExcessiveUseOfCobblestone** [comment]
no need for a list when pulled/pulledby can both only be single entities, just a single boolean that indicates can this shock transfer should be sufficient.

**Comment by kriskog** [comment]
> 
> 
> > Maybe the shock needs a list of already shocked things (EDIT: or excluded things, like the mob holding the defib) passed around as an arg then. Then you could also handle shocks up the chain (ie if you get shocked and you're being pulled, shock them too) without creating an infinite loop.
> 
> This sounds like the best solution, and I think that this would solve the "simplemob suiciding into a grill" issue that's keeping us from letting simplemobs join in on the fun.

I don't want simplemobs to be part of this. Most of them are hard to heal, have very little health and do not get stunned by shocks. 
All the reasons a list were suggested have also been sorted out in other ways.

**Comment by ExcessiveUseOfCobblestone** [comment]
sorry I got lazy reviewing this only for major optimizations which require a second review but me will get to it

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: source can never be pulling if it's a carbon since source is typed as an object

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: am I missing something? why is human/carbon separate checks but do the same?

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: source can never be pulledby if it's a carbon since source is typed as an object

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/carbon_defense.dm]: the correct fix is to retype source me thinks

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: same as above

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (kriskog) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Humans have their own electrocute_act, which calls the carbon_defense electrocute act at the end. It's just me faffing about with casting, I never did quite manage to fully wrap my head around casting. It seems to work fine when casting a carbon to human, so I'll remove the extra lines.

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (kriskog) [reply code/modules/mob/living/carbon/carbon_defense.dm]: They are both a reference to the same memory adress, I assume there is some implicit type casting involved at compiler level. Either way it works as expected.
Do you want me to explicitly cast these two vars to an object?

Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/carbon_defense.dm]: I want source to not be typed as an object (L226) if source can be things other than an object.

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: If you are checking for carbon then you type it as carbon

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/carbon/carbon_defense.dm]: If you are checking for carbon then you type it as carbon

**Review by ExcessiveUseOfCobblestone — DISMISSED** [review]

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Humans will still use their own electrocute_act so don't do anything special for them unless you have to call a var/proc specific to humans.

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (kriskog) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Right, yeah that makes sense. Source is only ever used for chat output anyway, before now.

File: code/modules/mob/living/carbon/carbon_defense.dm
Comment (kriskog) [reply code/modules/mob/living/carbon/carbon_defense.dm]: You.. are right, they do. something something wrapping head around type casting grumble grumble.

**Comment by ExcessiveUseOfCobblestone** [comment]
Theorycrafting here:

Say we have 3 people with 0 being not shocked and integers being shock iterations.

A0 - B0 - C0

A pulls B who pulls C who pulls A (why they're doing this I don't know, they can't move or they break the chain so they're trying to break the game I guess).

A gets shocked. A1 > B0 - C0  

**then** A0 x B2 > C0 **then** B0 x C3 > A0 **then** C0 x A4  > B0 **then** A0 x B5 > C0 **>>** ...

> is transfers shock to, x is WOULD transfer shock to but lol'd by source argument.

It becomes infinite loop, is there anything that protects against that? Is this even possible lol?

**Comment by kriskog** [comment]
Won't ever become infinite loop, since there is a reduction in damage (*.75) every time it propagates. On a roundstart power setup which does 30 damage it will jump 12 times (in either direction) before the shock damage is too low to register.  Would any shocks that doesn't stun be significantly higher than that?
If this is something you are really worried about, I would probably recommend we don't propagate shocks that do not stun, by just adding && stun to the checks. Other options are adding a stop_pulling() when shocked, or limiting how often a mob can be shocked.

Note however, even if it did cause an infinite loop it would runtime (maximum recursion level reached), and kill the proc.

**Comment by ExcessiveUseOfCobblestone** [comment]
ok good, I just needed to talk it out.

**Review by ExcessiveUseOfCobblestone — APPROVED** [review]

**Comment by optimumtact** [comment]
STOP APPROVING THINGS JUST MERGE THEM

**Comment by PKPenguin321** [comment]
Why don't u do it if you're going to comment about it anyways smh

**Comment by optimumtact** [comment]
because I shouldn't have to be the one who has to merge every pr

**Comment by ExcessiveUseOfCobblestone** [comment]
what did she mean by this...

**Pull request merged**
